common/clock/ratelimiter_test.go

```diff
@@ -0,0 +1,434 @@
+		if testDeadline.Before(deadline) {
+			deadline = testDeadline
+		}
+	}
+	for testnum := 0; !t.Failed() && time.Now().Before(deadline); testnum++ {
+		t.Run(fmt.Sprintf("attempt %v", testnum), func(t *testing.T) {
+			// parallel saves a fair bit of time but introduces a lot of CPU contention
+			// and that leads to a moderate amount of flakiness.
+			//
+			// unfortunately not recommended here.
+			// t.Parallel()
+
+			seed := time.Now().UnixNano()
+			seed = 1716164357569243000 // override seed to test a failing scenario
+			rng := rand.New(rand.NewSource(seed))
+			t.Logf("rand seed: %v", seed)
+
+			burst := rng.Intn(maxBurst)               // zero is uninteresting but allowed
+			limit := rate.Every(granularity * events) // refresh after each full round
+			t.Logf("limit: %v, burst: %v", granularity, burst)
+			initialNow := time.Now()
+			ts := NewMockedTimeSourceAt(initialNow)
+			compressedTS := NewMockedTimeSourceAt(initialNow)
+
+			actual := rate.NewLimiter(limit, burst)
+			wrapped := NewRatelimiter(limit, burst)
+			mocked := NewMockRatelimiter(ts, limit, burst)
+			compressed := NewMockRatelimiter(compressedTS, limit, burst)
+
+			// generate some non-colliding "1 == perform an Allow call" rounds.
+			calls := [rounds][events]int{} // using int just to print better than bool
+			for i := 0; i < len(calls[0]); i++ {
+				// pick one to set to true, or none (1 chance for none)
+				set := rng.Intn(len(calls) + 1)
+				if set < len(calls) {
+					calls[set][i] = 1
+				}
+			}
+			t.Log("round setup:")
+			for _, round := range calls {
+				t.Logf("\t%v", round)
+			}
+			/*
+				rounds look like:
+					[1 0 0 1 0 ...]
+					[0 1 1 0 0 ...]
+				which means that the first round will:
+					- call
+					- do nothing
+					- do nothing
+					- call ...
+				and by the end of that first array it'll reach the end of
+				the rate.Every time, and will begin recovering tokens during
+				the next array's runtime.
+				so the second round will:
+					- refresh 1 token while waiting
+					- consume 1 call
+					- consume 1 call
+					- refresh 1
+					- wait, no tokens refresh because none were consumed here last round
+
+				the `1`s do not overlap to avoid triggering a race between the
+				old call being refreshed and the new call consuming, as these
+				make for very flaky tests unless time is mocked.
+			*/
+
+			// record "to be executed" closures on the time-compressed ratelimiter too,
+			// so it can be checked at a sped up rate.
+			compressedReplay := [rounds][events]func(t *testing.T){}
+
+			ticker := time.NewTicker(granularity)
+			for round := range calls {
+				round := round // for closure
+				for event := range calls[round] {
+					event := event // for closure
+					<-ticker.C
+					ts.Advance(granularity)
+					debug(t, "Tokens before round, real: %0.2f, wrapped: %0.2f, mocked: %0.2f", actual.Tokens(), wrapped.Tokens(), mocked.Tokens())
+
+					if calls[round][event] == 1 {
```

Comment:
indentation goes too deep. `continue` early to reduce indentation.

Reply:
yep, totally agreed.  I kept growing this as I figured out more things to add, and I haven't gone back to clean it up yet.